### PR TITLE
fix Incorrect timeout error when using FindStringMatch()

### PR DESCRIPTION
**Pull request by Brinenas**
fix Timeout Error issue: when the fastclock stop running, then next matches will get a wrong fast.current and daeline, thus trigger an incorrect timeout error.

### CODE REVIEW

**Review by ccoVeille — CHANGES_REQUESTED** [review]
LGTM 👍

but I would expect some tests to be provided with the fix. It would avoid errors in any future refactoring

**Comment by Brinenas** [comment]
Add a new UT, If you comment the add code, the ut will failed.

File: fastclock_test.go [hunk 1]
```diff
@@ -56,3 +56,29 @@ func TestStopTimeoutClock(t *testing.T) {
...
+	// make the error time much bigger
+	time.Sleep(10 * clockPeriod)
+	nowTick := durationToTicks(time.Since(fast.start))
+	// before fix, fast.current will be the time fast stopped, and end is incorrect too
+	// after fix, fast.current will be current time.
+	d := makeDeadline(timeout)
+	gotTick := fast.current.read()
+	t.Logf("nowTick: %+v, gotTick: %+v", nowTick, gotTick)
+	if nowTick > gotTick {
+		t.Errorf("Expectd current should bigger than %v, got %v", gotTick, nowTick)
```
Comment (ccoVeille) [inline fastclock_test.go:78]: Nitpicking, maybe this

```suggestion
		t.Errorf("Expectd current should greater than %v, got %v", gotTick, nowTick)
```

**Review by ccoVeille — APPROVED** [review]

**Review by ccoVeille — APPROVED** [review]

**Comment by dlclark** [comment]
Thanks for submitting this. Can you add an integration-level test showing the behavior fix from the API consumer perspective?

**Comment by Brinenas** [comment]
Add a new UT for my own case. Maybe you guys can add a more elegant test.

**Comment by dlclark** [comment]
@Brinenas a couple issues I see:
* `fast.running` and `fast.start` are protected by `fast.mu` for read/write, so we can't access them without getting the lock
* I don't think we want to get that lock every time we need to make a deadline because this is an edge case. I've refactored the code to only do this if we already know the deadline needs to be extended.

I did a couple other small tweaks to the test as well. Let me know if you see any issues with my change.

**Comment by bradrevans** [comment]
This change fixed a similar issue for me - intermittent false timeouts seemingly triggered by a large volume of sequential regex matching that appeared after the inclusion of the fast clock back in v1.8.0. Appreciate theres probably a lot of other stuff on your plate but was wondering when we might be able to expect a new tagged release including this fix? Thanks for the effort on resolving this! <3

**Comment by dlclark** [comment]
@bradrevans this is done.

**Comment by bradrevans** [comment]
> @bradrevans this is done.

Ah. Thanks!
